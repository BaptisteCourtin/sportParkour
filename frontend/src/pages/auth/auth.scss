.auth {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  width: 100%;

  .imageTop {
    background: url("https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSCZlf5lc5tX-0gY-y94pGS0mQdL-D0lCH2OQ&s");
    background-position: center;
    background-size: cover;
    background-repeat: no-repeat;

    display: flex;
    flex-direction: column;
    align-items: center;

    width: 100%;
    height: 200px;
    border-radius: var(--radius);

    padding: 70px var(--espace);

    h1 {
      color: var(--text1);
    }
  }

  form {
    width: calc(100% - 40px);
    max-width: 700px;
    background-color: var(--middle);

    padding: var(--espace);
    border-radius: var(--radius);
    border-top: var(--line) solid var(--blue);

    transform: translateY(-50px);

    .topForm {
      display: flex;
      flex-direction: column;

      h2 {
        color: var(--blue);
        align-self: flex-start;

        &::before {
          content: "";
          width: var(--losange);
          height: var(--losange);
          background-color: var(--blue);
          position: absolute;
          transform: rotate(45deg);
          top: 26px;
          left: -7.5px;
        }
      }
    }

    .inscr {
      align-self: flex-end;
      text-decoration: underline;
      font-size: var(--fs-mini);
    }

    .oublie {
      align-self: flex-start;
      text-decoration: underline;
      font-size: var(--fs-mini);
    }

    .legale {
      display: flex;
      flex-direction: column;
      text-decoration: underline;
      font-size: var(--fs-mini);
    }

    button {
      align-self: flex-end;
    }
  }
}
