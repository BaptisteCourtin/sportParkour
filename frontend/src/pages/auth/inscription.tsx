import React, { useState } from "react";
import Link from "next/link";
import { useRouter } from "next/router";
import { useForm } from "react-hook-form";
import { object, ref, string } from "yup";
import { yupResolver } from "@hookform/resolvers/yup";

import {
  UserInputRegisterEntity,
  useInscriptionMutation,
} from "@/types/graphql";

import TextField from "@mui/material/TextField";
import FormControlLabel from "@mui/material/FormControlLabel";
import Checkbox from "@mui/material/Checkbox";
import InputAdornment from "@mui/material/InputAdornment";
import IconButton from "@mui/material/IconButton";

import { toast } from "react-hot-toast";
import { FaEye } from "react-icons/fa6";
import { FaEyeSlash } from "react-icons/fa6";
import { FaArrowRight } from "react-icons/fa6";

import SearchBarCommuneName from "@/components/user/searchBarCommuneName";

import {
  LENGTH_EMAIL,
  LENGTH_MIN_PASSWORD,
  LENGTH_NOM,
  LENGTH_MAX_PASSWORD,
  LENGTH_PHONE,
} from "../../../../variablesLength";

let authSchema = object({
  email: string()
    .max(LENGTH_EMAIL)
    .email("Votre email doit être valide")
    .required("Veuillez entrer votre email"),
  password: string()
    .min(
      LENGTH_MIN_PASSWORD,
      `Utilisez un mot de passe avec au moins ${LENGTH_MIN_PASSWORD} caractères`
    )
    .max(
      LENGTH_MAX_PASSWORD,
      `Utilisez un mot de passe avec au maximum ${LENGTH_MAX_PASSWORD} caractères`
    )
    .matches(/[A-Z]/, "Utilisez au moins une majuscule")
    .matches(/[a-z]/, "Utilisez au moins une minuscule")
    .matches(/[0-9]/, "Utilisez au moins un chiffre")
    .matches(/[!@#$%^&*(),.?":{}|<>]/, "Utilisez au moins un caractère spécial")
    .required("Veuillez entrer votre mot de passe"),
  password2: string()
    .oneOf([ref("password")], "Les mots de passe ne correspondent pas")
    .required("Veuillez confirmer votre mot de passe"),

  name: string().max(LENGTH_NOM).required("Veuillez entrer votre nom"),
  firstname: string().max(LENGTH_NOM).required("Veuillez entrer votre prénom"),

  phone: string()
    .max(LENGTH_PHONE, "tapez votre numéro sans espace et sans le +33")
    .test(
      "len",
      "tapez les 10 chiffres de votre numéro, sans espace et sans le +33",
      (val) => {
        if (val == undefined) {
          return true;
        }
        return val.length == 0 || val.length == LENGTH_PHONE;
      }
    ),
});

interface UserInputRegisterWith2PasswordsEntity
  extends UserInputRegisterEntity {
  password2: string;
}

// -----------------------------------------------------------------------------------

const inscription = () => {
  const router = useRouter();

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm({
    resolver: yupResolver(authSchema),
  });

  const [inscription, { data, loading, error }] = useInscriptionMutation();

  const handleInscription = (
    dataForm: UserInputRegisterWith2PasswordsEntity
  ): void => {
    const { password2, codePostal, ...data } = dataForm;
    const infos = {
      city: selectedCommuneName,
      codePostal: selectedCommuneCodePostal,
      ...data,
    };

    inscription({
      variables: { infos: infos },
      onCompleted(data) {
        if (data.inscription.success) {
          toast.success(data.inscription.message);
          router.push("/");
        }
      },
      onError(error) {
        toast.error(error.message);
      },
    });
  };

  // --- DEAL WITH LENGTH DURING MODIF ---
  const [values, setValues] = useState({
    firstname: "",
    name: "",
    phone: "",
    email: "",
    password: "",
    password2: "",
  });

  const handleChangeAThing = (name: string, value: any) => {
    setValues({ ...values, [name]: value });
  };

  // --- SEE PASSWORDS ---
  const [showPassword, setShowPassword] = useState(false);
  const [showPassword2, setShowPassword2] = useState(false);

  // --- API COMMUNES ---
  const [selectedCommuneName, setSelectedCommuneName] = useState("");
  const [selectedCommuneCodePostal, setSelectedCommuneCodePostal] =
    useState("");

  return (
    <main className="auth">
      <div className="imageTop">
        <h1>POSSEDEZ UN COMPTE PARKOUR</h1>
      </div>

      <form onSubmit={handleSubmit(handleInscription)} className="bigForm">
        <div className="topForm">
          <h2>INSCRIPTION</h2>
          <Link className="inscr" href="/auth/login">
            Vous avez déjà un compte ?
          </Link>
        </div>

        <div className="champ">
          <TextField
            className="mui-input"
            fullWidth
            variant="outlined"
            label="Votre prénom"
            required
            {...register("firstname")}
            id="firstname"
            name="firstname"
            type="firstname"
            inputProps={{ maxLength: LENGTH_NOM }}
            onChange={(e) => handleChangeAThing("firstname", e.target.value)}
          />
          <span>
            {values.firstname.length > 0
              ? `${values.firstname.length}/${LENGTH_NOM}`
              : ""}
          </span>
          <p className="error">{errors?.firstname?.message}</p>
        </div>
        <div className="champ">
          <TextField
            className="mui-input"
            fullWidth
            variant="outlined"
            label="Votre nom"
            required
            {...register("name")}
            id="name"
            name="name"
            type="text"
            inputProps={{ maxLength: LENGTH_NOM }}
            onChange={(e) => handleChangeAThing("name", e.target.value)}
          />
          <span>
            {values.name.length > 0
              ? `${values.name.length}/${LENGTH_NOM}`
              : ""}
          </span>
          <p className="error">{errors?.name?.message}</p>
        </div>

        <div className="containerMiniChamp">
          <div className="champ">
            <SearchBarCommuneName
              userValue={selectedCommuneName}
              setSelectedCommuneName={setSelectedCommuneName}
              setSelectedCommuneCodePostal={setSelectedCommuneCodePostal}
            />
          </div>

          <div className="champ">
            <TextField
              className="mui-input codePostal"
              fullWidth
              variant="outlined"
              label="Votre code postal"
              value={selectedCommuneCodePostal}
              id="codePostal"
              name="codePostal"
              type="text"
              InputProps={{
                readOnly: true,
              }}
            />
          </div>
        </div>

        <div className="champ">
          <TextField
            className="mui-input"
            fullWidth
            variant="outlined"
            label="Votre numéro de téléphone"
            {...register("phone")}
            id="phone"
            name="phone"
            type="text"
            inputProps={{ maxLength: LENGTH_PHONE }}
            onChange={(e) => handleChangeAThing("phone", e.target.value)}
          />
          <span>
            {values.phone.length > 0
              ? `${values.phone.length}/${LENGTH_PHONE}`
              : ""}
          </span>
          <p className="error">{errors?.phone?.message}</p>
        </div>

        <hr />

        <div className="champ">
          <TextField
            className="mui-input"
            fullWidth
            variant="outlined"
            label="Votre email"
            required
            {...register("email")}
            id="email"
            name="email"
            type="text"
            inputProps={{ maxLength: LENGTH_EMAIL }}
            onChange={(e) => handleChangeAThing("email", e.target.value)}
          />
          <span>
            {values.email.length > 0
              ? `${values.email.length}/${LENGTH_EMAIL}`
              : ""}
          </span>
          <p className="error">{errors?.email?.message}</p>
        </div>
        <div className="champ">
          <TextField
            className="mui-input"
            fullWidth
            variant="outlined"
            label="Votre mot de passe"
            required
            {...register("password")}
            id="password"
            name="password"
            type={showPassword ? "text" : "password"}
            inputProps={{ maxLength: LENGTH_MAX_PASSWORD }}
            onChange={(e) => handleChangeAThing("password", e.target.value)}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end" className="buttonEye">
                  <IconButton onClick={() => setShowPassword(!showPassword)}>
                    {showPassword ? <FaEyeSlash /> : <FaEye />}
                  </IconButton>
                </InputAdornment>
              ),
            }}
          />
          <span>
            {values.password.length > 0
              ? `${values.password.length}/${LENGTH_MAX_PASSWORD}`
              : ""}
          </span>
          <p className="error">{errors?.password?.message}</p>
        </div>
        <div className="champ">
          <TextField
            className="mui-input"
            fullWidth
            variant="outlined"
            label="Réécrivez votre mot de passe"
            required
            {...register("password2")}
            id="password2"
            name="password2"
            type={showPassword2 ? "text" : "password"}
            inputProps={{ maxLength: LENGTH_MAX_PASSWORD }}
            onChange={(e) => handleChangeAThing("password2", e.target.value)}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end" className="buttonEye">
                  <IconButton onClick={() => setShowPassword2(!showPassword2)}>
                    {showPassword2 ? <FaEyeSlash /> : <FaEye />}
                  </IconButton>
                </InputAdornment>
              ),
            }}
          />
          <span>
            {values.password2.length > 0
              ? `${values.password2.length}/${LENGTH_MAX_PASSWORD}`
              : ""}
          </span>
          <p className="error">{errors?.password2?.message}</p>
        </div>

        <FormControlLabel
          required
          control={<Checkbox />}
          label="j'accepte les CGU"
          data-testid="accept-cgu-checkbox"
        />
        <div className="legale">
          <Link href="/infos/cgu">Voir les CGU</Link>
          <Link href="/infos/mentionsLegales">Voir les mentions légales</Link>
          <Link href="/infos/politiqueDeConfidentialite">
            Voir la politique de confidentialité
          </Link>
        </div>

        <button type="submit" disabled={loading}>
          SUIVANT <FaArrowRight />
        </button>

        <div>
          <span>{error?.message}</span>
        </div>
      </form>
    </main>
  );
};

export default inscription;
