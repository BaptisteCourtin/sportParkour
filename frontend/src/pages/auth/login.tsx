import React, { useState } from "react";
import Link from "next/link";
import { useRouter } from "next/router";
import {
  UserInputAuthEntity,
  useAuthentificationLazyQuery,
} from "@/types/graphql";

import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import { object, string } from "yup";

import { toast } from "react-hot-toast";
import TextField from "@mui/material/TextField";
import { FaArrowRight } from "react-icons/fa6";
import InputAdornment from "@mui/material/InputAdornment";
import IconButton from "@mui/material/IconButton";
import { FaEye } from "react-icons/fa6";
import { FaEyeSlash } from "react-icons/fa6";

let loginSchema = object({
  email: string()
    .email("Votre email doit être valide")
    .required("Veuillez entrer votre email"),
  password: string()
    // .min(12, "Votre mot de passe fait au moins 12 caractères") // pour que je continue à utiliser 0000
    .max(100, "Votre mot de passe fait moins de 100 caractères")
    .required("Veuillez entrer votre mot de passe"),
});

// mettre le message de success en snackbar
const login = () => {
  const router = useRouter();

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm({
    resolver: yupResolver(loginSchema),
  });

  const [authentification, { data, loading, error }] =
    useAuthentificationLazyQuery({ fetchPolicy: "no-cache" });

  const handleAuthentification = (dataForm: UserInputAuthEntity): void => {
    if (dataForm.email && dataForm.password) {
      authentification({
        variables: { infos: dataForm },
        onCompleted(data) {
          if (data.authentification.success) {
            toast.success(data.authentification.message);
            router.push("/");
          } else {
            toast.success(data.authentification.message);
          }
        },
        onError(error) {
          toast.error(error.message);
        },
      });
    }
  };

  // --- SEE PASSWORDS ---
  const [showPassword, setShowPassword] = useState(false);

  return (
    <main className="auth">
      <div className="imageTop">
        <h1>UTILISEZ VOTRE ESPACE PERSONNEL</h1>
      </div>

      <form onSubmit={handleSubmit(handleAuthentification)}>
        <div className="topForm">
          <h2>CONNECTEZ VOUS</h2>
          <Link className="inscr" href="/auth/inscription">
            Vous voulez un compte ?
          </Link>
        </div>
        <div className="champ">
          <TextField
            className="mui-input"
            fullWidth
            variant="outlined"
            label="Votre email"
            required
            {...register("email")}
            id="email"
            type="text"
            name="email"
            inputProps={{ maxLength: 255 }}
          />
          <p className="error">{errors?.email?.message}</p>
        </div>

        <div className="champ">
          <TextField
            className="mui-input"
            fullWidth
            variant="outlined"
            label="Votre mot de passe"
            required
            {...register("password")}
            id="password"
            type={showPassword ? "text" : "password"}
            name="password"
            inputProps={{ maxLength: 100 }}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">
                  <IconButton onClick={() => setShowPassword(!showPassword)}>
                    {showPassword ? <FaEyeSlash /> : <FaEye />}
                  </IconButton>
                </InputAdornment>
              ),
            }}
          />
          <p className="error">{errors?.password?.message}</p>
        </div>

        <Link className="oublie" href="/auth/resetPassword/resetPassword">
          Mot de passe oublié ?
        </Link>

        <button type="submit" disabled={loading}>
          SUIVANT <FaArrowRight />
        </button>

        <div>
          <span>{error?.message}</span>
        </div>
      </form>
    </main>
  );
};

export default login;
