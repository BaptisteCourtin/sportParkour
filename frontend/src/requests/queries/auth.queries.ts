import { gql } from "@apollo/client";

export const AUTHENTIFICATION = gql`
  query Authentification($infos: UserInputAuthEntity!) {
    authentification(infos: $infos) {
      message
      success
    }
  }
`;

// {
//   "infos": {
//     "password": "oui",
//     "email": "oui@oui.com"
//   }
// }

export const LOGOUT = gql`
  query Logout {
    logout {
      message
      success
    }
  }
`;
