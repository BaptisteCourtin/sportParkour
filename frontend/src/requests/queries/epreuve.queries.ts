import { gql } from "@apollo/client";

export const GET_EPREUVE = gql`
  query GetEpreuve($getEpreuveId: Float!) {
    getEpreuve(id: $getEpreuveId) {
      id
      title
      description
      easyToDo
      mediumToDo
      hardToDo
      videoLink
      images {
        id
        lien
      }
    }
  }
`;

export const GET_LIST_EPREUVES = gql`
  query GetListEpreuves {
    getListEpreuve {
      id
      title
    }
  }
`;

export const GET_LIST_EPREUVES_BY_SEARCH = gql`
  query GetListEpreuvesBySearch($search: String) {
    getListBySearch(search: $search) {
      id
      title
    }
  }
`;
