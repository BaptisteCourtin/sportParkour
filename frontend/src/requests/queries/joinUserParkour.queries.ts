import { gql } from "@apollo/client";

export const GET_FAV_BY_EMAIL = gql`
  query GetUserFavByToken {
    getUserFavByToken {
      parkour_id
      note
      favoris
      parkours {
        id
        title
      }
    }
  }
`;
