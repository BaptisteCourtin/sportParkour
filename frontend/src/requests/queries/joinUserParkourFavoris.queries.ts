import { gql } from "@apollo/client";

export const GET_FAV_BY_TOKEN_AND_ID_PARKOUR = gql`
  query GetUserFavByTokenAndParkourId($parkourId: Float!) {
    getUserFavByTokenAndParkourId(parkourId: $parkourId)
  }
`;

// {
//   "parkourId": 3
// }

export const GET_ALL_FAV_BY_TOKEN = gql`
  query GetAllUserFavByToken {
    getAllUserFavByToken {
      parkour {
        id
        title
        time
        length
        difficulty
        city
        note
        nbVote
        images {
          id
          lien
        }
        epreuves {
          id
          title
        }
      }
    }
  }
`;
