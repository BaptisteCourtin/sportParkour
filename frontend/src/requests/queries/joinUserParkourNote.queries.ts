import { gql } from "@apollo/client";

export const GET_NOTE_BY_TOKEN_AND_ID_PARKOUR = gql`
  query GetUserNoteByTokenAndParkourId($parkourId: Float!) {
    getUserNoteByTokenAndParkourId(parkourId: $parkourId) {
      note
      commentaire
    }
  }
`;

// {
//   "parkourId": 3
// }

export const GET__ALL_NOTE_BY_TOKEN = gql`
  query GetAllUserNoteByToken {
    getAllUserNoteByToken {
      note
      commentaire
      parkour {
        id
        title
        time
        length
        difficulty
        city
        note
        nbVote
        images {
          id
          lien
        }
        epreuves {
          id
          title
        }
      }
    }
  }
`;
