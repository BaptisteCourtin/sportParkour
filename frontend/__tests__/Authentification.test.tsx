import Login from "@/pages/auth/login";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { AUTHENTIFICATION } from "@/requests/queries/auth.queries";
import { MemoryRouterProvider } from "next-router-mock/MemoryRouterProvider";
import { MockedProvider } from "@apollo/client/testing";

jest.mock("next/router", () => ({
  ...jest.requireActual("next-router-mock"),
  push: jest.fn(),
}));

const mocksAuthentification = [
  {
    request: {
      query: AUTHENTIFICATION,
      variables: {
        infos: { email: "kevin75du75@gmail.com", password: "0000" },
      },
    },
    result: {
      data: {
        authentification: {
          message: "Salut! Viens parkourir le monde !",
          success: true,
        },
      },
    },
  },
];

it("Snapshot du login", async () => {
  const { container, getByText, getByPlaceholderText } = render(
    <MemoryRouterProvider url="/">
      <MockedProvider mocks={mocksAuthentification} addTypename={false}>
        <Login />
      </MockedProvider>
    </MemoryRouterProvider>
  );

  // Remplissage du formulaire avec des données de test
  fireEvent.change(getByPlaceholderText("Indiquez votre email"), {
    target: { value: "kevin75du75@gmail.com" },
  });
  fireEvent.change(getByPlaceholderText("Indiquez votre mot de passe"), {
    target: { value: "0000" },
  });

  // Soumission du formulaire
  fireEvent.submit(getByText("Se connecter"));

  // On attend un certain temps pour permettre à onCompleted de s'exécuter
  await waitFor(() => {}, { timeout: 1000 });

  // On vérifie le snapshot après que onCompleted a eu le temps de s'exécuter
  await waitFor(
    () => {
      expect(container).toMatchSnapshot();
    },
    { timeout: 10000 }
  );
});
