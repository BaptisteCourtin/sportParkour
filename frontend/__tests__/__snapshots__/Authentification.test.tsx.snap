// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot du login 1`] = `
<div>
  <main
    class="auth"
  >
    <a
      href="/auth/inscription"
    >
      Inscription
    </a>
    <form>
      <h1>
        Authentification
      </h1>
      <div>
        <label
          for="email"
        >
          Votre Email
        </label>
        <input
          id="email"
          name="email"
          placeholder="Indiquez votre email"
          type="text"
        />
        <p
          class="error"
        />
      </div>
      <div>
        <label
          for="password"
        >
          Votre Password
        </label>
        <input
          id="password"
          name="password"
          placeholder="Indiquez votre mot de passe"
          type="password"
        />
        <p
          class="error"
        />
      </div>
      <button
        disabled=""
        type="submit"
      >
        Se connecter
      </button>
      <div>
        <span />
        <span />
      </div>
    </form>
  </main>
</div>
`;
